src/vs/workbench/services/notebook/common/notebookDocumentService.ts

```diff
@@ -0,0 +1,91 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { VSBuffer, decodeBase64, encodeBase64 } from 'vs/base/common/buffer';
+import { ResourceMap } from 'vs/base/common/map';
+import { Schemas } from 'vs/base/common/network';
+import { URI } from 'vs/base/common/uri';
+import { InstantiationType, registerSingleton } from 'vs/platform/instantiation/common/extensions';
+import { createDecorator } from 'vs/platform/instantiation/common/instantiation';
+
+export const INotebookDocumentService = createDecorator<INotebookDocumentService>('notebookDocumentService');
+
+export interface INotebookDocument {
+	readonly uri: URI;
+	getCellIndex(cellUri: URI): number | undefined;
+}
+
+const _lengths = ['W', 'X', 'Y', 'Z', 'a', 'b', 'c', 'd', 'e', 'f'];
+const _padRegexp = new RegExp(`^[${_lengths.join('')}]+`);
+const _radix = 7;
+export function parse(cell: URI): { notebook: URI; handle: number } | undefined {
+	if (cell.scheme !== Schemas.vscodeNotebookCell) {
+		return undefined;
+	}
+
+	const idx = cell.fragment.indexOf('s');
+	if (idx < 0) {
+		return undefined;
+	}
+
+	const handle = parseInt(cell.fragment.substring(0, idx).replace(_padRegexp, ''), _radix);
+	const _scheme = decodeBase64(cell.fragment.substring(idx + 1)).toString();
+
+	if (isNaN(handle)) {
+		return undefined;
+	}
+	return {
+		handle,
+		notebook: cell.with({ scheme: _scheme, fragment: null })
+	};
+}
+
+export function generate(notebook: URI, handle: number): URI {
+
+	const s = handle.toString(_radix);
+	const p = s.length < _lengths.length ? _lengths[s.length - 1] : 'z';
+
+	const fragment = `${p}${s}s${encodeBase64(VSBuffer.fromString(notebook.scheme), true, true)}`;
+	return notebook.with({ scheme: Schemas.vscodeNotebookCell, fragment });
+}
+
+export interface INotebookDocumentService {
+	readonly _serviceBrand: undefined;
+
+	getNotebook(uri: URI): INotebookDocument | undefined;
+	addNotebookDocument(document: INotebookDocument): void;
+	removeNotebookDocument(document: INotebookDocument): void;
+}
+
+export class NotebookDocumentWorkbenchService implements INotebookDocumentService {
```

Comment:
Curious: rather than introducing an entire new service for this, could we make this work via the `ILabelService` that already allows to register formatters for a scheme?

Reply:
@bpasero explored this and based on my understanding, the formatter is declarative but doesn't support generate formatted uri name/description on the fly (Uri for notebook cells are some sort of anonymous so we can't parse the info from it statically).

Regarding new service, we start to have more components (e.g., Search) depend on the notebook contrib for reading notebook document related info. It's probably a good time to bump the notebook document service up to the service layer (conceptually it also makes good sense as we see notebook as first party), I'll gradually make `notebookservice` itself slimmer and internal to the notebook contrib itself.

Reply:
Yeah I was not sure if the existing label service already provides enough and I am easy to make it support this scenario but if we are building more need for a notebook service in core I do not mind much either.